### ISSUE TITLE
unpack 'a3#a*' misbehaving

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#389]
Migrated from [rt.perl.org#1224](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=1224) (status was 'resolved')

Searchable as RT1224$

**Comment by p5pRT** [comment]
### From ilya@math.ohio-state.edu

&nbsp;      DB\<1> x unpack 'Z3\#a\*'\, pack 'Z3\#a\*'\, '123456789012345'
&nbsp;    0  123456789012345

OK\.

&nbsp;      DB\<2> x pack 'a3\#a\*'\, '123456789012345'
&nbsp;    0  "15\\c@&#8203;123456789012345"

OK\.

&nbsp;      DB\<3> x unpack 'a3\#a\*'\, pack 'a3\#a\*'\, '123456789012345'
&nbsp;    Argument "15" isn't numeric in unpack at \(eval 7\) line 2\, \<IN> line 3\.
&nbsp;        eval '\($@&#8203;\, $\!\, $^E\, $\,\, $/\, $\\\\\, $^W\) = @&#8203;saved;package main; $^D = $^D | $DB&#8203;::db\_stop;
&nbsp;      unpack \\'a3\#a\*\\'\, pack \\'a3\#a\*\\'\, \\'123456789012345\\';

&nbsp;    ;' called at lib/perl5db\.pl line 1246
&nbsp;        DB&#8203;::eval called at lib/perl5db\.pl line 1140
&nbsp;        DB&#8203;::DB called at \-e line 1
&nbsp;    0  ''

The warning is understandable\, but it is only a warning\!  As you see\,
the result is empty\, not 123456789012345 as expected\.

Ilya



---
Site configuration information for perl 5\.00560&#8203;:

Configured by ilya at Wed Aug  4 22&#8203;:44&#8203;:30 EDT 1999\.

Summary of my perl5 \(revision 5\.0 version 5 subversion 60\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.7\, archname=sun4\-solaris
&nbsp;    uname='sunos monk 5\.7 generic\_106541\-05 sun4u sparc '
&nbsp;    config\_args='\-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef useperlio=undef d\_sfio=undef
&nbsp;    use64bits=undef usemultiplicity=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, optimize='\-O'\, gccversion=
&nbsp;    cppflags='\-I/opt/local/include \-I/opt/gnu/include'
&nbsp;    ccflags ='\-I/opt/local/include \-I/opt/gnu/include'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    alignbytes=8\, usemymalloc=y\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/opt/local/lib \-L/opt/gnu/lib'
&nbsp;    libpth=/opt/local/lib /opt/gnu/lib /lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-lgdbm \-ldl \-lm \-lrt \-lc \-lcrypt \-lsec
&nbsp;    libc=/lib/libc\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-KPIC'\, lddlflags='\-G \-L/opt/local/lib \-L/opt/gnu/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.00560&#8203;:
&nbsp;    lib
&nbsp;    /home/ilya/perl
&nbsp;    /home/ilya/perl/lib/perl5
&nbsp;    /home/ilya/perl/lib/perl5/site\_perl
&nbsp;    /home/ilya/perl/lib/perl5/site\_perl/sun4\-solaris
&nbsp;    /home/ilya/perl/lib/perl5/site\_perl/5\.005/sun4\-solaris
&nbsp;    /home/ilya/perl/lib/perl5/5\.00503/
&nbsp;    /opt/lib/perl5/5\.00560/sun4\-solaris
&nbsp;    /opt/lib/perl5/5\.00560
&nbsp;    /opt/lib/site\_perl/5\.00560/sun4\-solaris
&nbsp;    /opt/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.00560&#8203;:
&nbsp;    HOME=/home/ilya
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH=/opt/local/lib&#8203;:/usr/openwin/lib&#8203;:/opt/local/lib/rvplayer5\.0&#8203;:/opt/X11/lib
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/ilya/bin&#8203;:/opt/SUNWspro/bin&#8203;:/usr/dt/bin&#8203;:/usr/ccs/bin&#8203;:/opt/java/bin&#8203;:/opt/netscape&#8203;:/opt/microsoft/bin&#8203;:/opt/gnu/bin&#8203;:/usr/openwin/bin&#8203;:/opt/X11/bin&#8203;:/opt/unsup/beta/bin&#8203;:/home/tex/bin&#8203;:/opt/local/bin&#8203;:/usr/bin&#8203;:/opt/local/games&#8203;:/opt/unsup/gnu/bin&#8203;:/opt/unsup/ilya/bin&#8203;:\.
&nbsp;    PERLLIB=/home/ilya/perl&#8203;:/home/ilya/perl/lib/perl5&#8203;:/home/ilya/perl/lib/perl5/site\_perl&#8203;:/home/ilya/perl/lib/perl5/site\_perl/sun4\-solaris&#8203;:/home/ilya/perl/lib/perl5/site\_perl/5\.005/sun4\-solaris&#8203;:/home/ilya/perl/lib/perl5/5\.00503/
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/usr/bin/tcsh

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Tue\, 17 Aug 1999 at 03&#8203;:37&#8203;:30 \-0400\, Ilya Zakharevich wrote&#8203;:

>       DB\<3> x unpack 'a3\#a\*'\, pack 'a3\#a\*'\, '123456789012345'
>     Argument "15" isn't numeric in unpack at \(eval 7\) line 2\, \<IN> line 3\.
>         eval '\($@&#8203;\, $\!\, $^E\, $\,\, $/\, $\\\\\, $^W\) = @&#8203;saved;package main; $^D = $^D | $DB&#8203;::db\_stop;
>       unpack \\'a3\#a\*\\'\, pack \\'a3\#a\*\\'\, \\'123456789012345\\';

\[and the unpack gets a zero figure for the length because of the
embedded null\]

I just used POPi to pick up the number\, and this calls Perl\_sv\_2iv\. This
seems to be less forgiving than other conversion routines\. Anyhow\,
looking at Perl\_sv\_2iv\, there's only one place I can see that makes it
emit the warning\, and\, sure enough\, it sets the returned value to zero\.
Most other circumstances \(I've not traced them\) seem to lead to the
warning being emitted\, but nevertheless a "best guess" being returned\.

So&#8203;: Am I misusing POPi \[pp\.c line 3399 or so\]\, or is there a bug in
Perl\_sv\_2iv?

And\, I can't help feeling that /\\d\+\\0\+/ looks like a bona\-fide number to
me\. Is the warning justified at all? There's no warning for /\\d\+\\s\+/\.

Ian

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Ian Phillipps \<ian@&#8203;dial\.pipex\.com> writes&#8203;:
>
>And\, I can't help feeling that /\\d\+\\0\+/ looks like a bona\-fide number to
>me\. Is the warning justified at all? There's no warning for /\\d\+\\s\+/\.

\\d\+\\0 looks \_more\_ like a number to me than \\d\+\\s \!

\-\- 
Nick Ing\-Simmons \<nik@&#8203;tiuk\.ti\.com>
Via\, but not speaking for&#8203;: Texas Instruments Ltd\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed\, Aug 18\, 1999 at 12&#8203;:47&#8203;:09PM \+0100\, Ian Phillipps wrote&#8203;:
> On Tue\, 17 Aug 1999 at 03&#8203;:37&#8203;:30 \-0400\, Ilya Zakharevich wrote&#8203;:
> 
> >       DB\<3> x unpack 'a3\#a\*'\, pack 'a3\#a\*'\, '123456789012345'
> >     Argument "15" isn't numeric in unpack at \(eval 7\) line 2\, \<IN> line 3\.
> >         eval '\($@&#8203;\, $\!\, $^E\, $\,\, $/\, $\\\\\, $^W\) = @&#8203;saved;package main; $^D = $^D | $DB&#8203;::db\_stop;
> >       unpack \\'a3\#a\*\\'\, pack \\'a3\#a\*\\'\, \\'123456789012345\\';
> 
> \[and the unpack gets a zero figure for the length because of the
> embedded null\]
> 
> I just used POPi to pick up the number\, and this calls Perl\_sv\_2iv\. This
> seems to be less forgiving than other conversion routines\.

What is/was strange to me is that POPi is used in pp\_add as well\, but
it gives a correct result&#8203;:

 > perl \-wle "$two = qq\(2\\0\); print 2\+$two"
 Argument "2" isn't numeric in add at \-e line 1\.
 4

Aha\!  It is not POPi\, it is POPn\!

&nbsp;  > perl \-Minteger \-wle "$two = qq\(2\\0\); print 2\+$two"
&nbsp;  Argument "2" isn't numeric in i\_add at \-e line 1\.
&nbsp;  2

So it is \*definitely\* a bug in sv\_2iv\.

Ilya

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed\, 18 Aug 1999 at 17&#8203;:50&#8203;:59 \+0100\, Nick Ing\-Simmons wrote&#8203;:
> Ian Phillipps \<ian@&#8203;dial\.pipex\.com> writes&#8203;:
> >
> >And\, I can't help feeling that /\\d\+\\0\+/ looks like a bona\-fide number to
> >me\. Is the warning justified at all? There's no warning for /\\d\+\\s\+/\.
> 
> \\d\+\\0 looks \_more\_ like a number to me than \\d\+\\s \!

I propose this\, which will make Ilya's case work OK\, but doesn't fix the
inconsistency with sv\_2iv\. This returns 0 if there's a warning\, unlike the
other sv\_2xv routines\, which run the conversion anyway\.

My patch will stop the looks\_like\_number scan at the first "\\0"\, which
is after all what atof\(\) does\.  Hence\, "123\\0foo" will now be converted
to a number without any warnings\.

None of the 1440 op/numconvert tests is affected by this patch :\-\)

\(My apologies if this doesn't go cleanly on to 60\.\.\.\)

<details><summary>Inline Patch</summary>

```diff
--- sv.c.orig	Mon Aug  2 22:29:12 1999
+++ sv.c	Thu Aug 19 12:31:18 1999
@@ -1635,5 +1635,5 @@
     while (isSPACE(*s))
 	s++;
-    if (s >= send)
+    if (s >= send || *s == 0)
 	return numtype;
     if (len == 10 && memEQ(sbegin, "0 but true", 10))

Ian
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu\, Aug 19\, 1999 at 12&#8203;:56&#8203;:25PM \+0100\, Ian Phillipps wrote&#8203;:
> > Ian Phillipps \<ian@&#8203;dial\.pipex\.com> writes&#8203;:
> > >
> > >And\, I can't help feeling that /\\d\+\\0\+/ looks like a bona\-fide number to
> > >me\. Is the warning justified at all? There's no warning for /\\d\+\\s\+/\.
> > 
> > \\d\+\\0 looks \_more\_ like a number to me than \\d\+\\s \!
> 
> I propose this\, which will make Ilya's case work OK\, but doesn't fix the
> inconsistency with sv\_2iv\. This returns 0 if there's a warning\, unlike the
> other sv\_2xv routines\, which run the conversion anyway\.
> 
> My patch will stop the looks\_like\_number scan at the first "\\0"\, which
> is after all what atof\(\) does\.  Hence\, "123\\0foo" will now be converted
> to a number without any warnings\.

I would prefer sv\_2iv\(\) call conversion routine after emitting a
warning\.  \(I see no value in the testcase I sent \- one can always use
'A' instead of 'a' in the unpack specifier\.\)

> None of the 1440 op/numconvert tests is affected by this patch :\-\)

Well\, they fence a different can of worms\.\.\.

Ilya

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu\, 19 Aug 1999 at 14&#8203;:32&#8203;:08 \-0400\, Ilya Zakharevich wrote&#8203;:
> On Thu\, Aug 19\, 1999 at 12&#8203;:56&#8203;:25PM \+0100\, Ian Phillipps wrote&#8203;:
> > > Ian Phillipps \<ian@&#8203;dial\.pipex\.com> writes&#8203;:

Ian 1> I can't help feeling that /\\d\+\\0\+/ looks like a bona\-fide number to
Ian 1> me\. Is the warning justified at all? There's no warning for /\\d\+\\s\+/\.

Nick > \\d\+\\0 looks \_more\_ like a number to me than \\d\+\\s \!

Ian 2 > I propose this\, which will make Ilya's case work OK\, but doesn't fix the
Ian 2 > inconsistency with sv\_2iv\. This returns 0 if there's a warning\, unlike the
Ian 2 > other sv\_2xv routines\, which run the conversion anyway\.

Ian 2 > My patch will stop the looks\_like\_number scan at the first "\\0"\, which
Ian 2 > is after all what atof\(\) does\.  Hence\, "123\\0foo" will now be converted
Ian 2 > to a number without any warnings\.

Ilya > I would prefer sv\_2iv\(\) call conversion routine after emitting a
Ilya > warning\.  \(I see no value in the testcase I sent \- one can always use
Ilya > 'A' instead of 'a' in the unpack specifier\.\)

This patchlet was in support of Nick's agreement with my suggestion that
"123\\0" should be treated as a fully legitimate number\. I still think
it's Good Thing and crave the attention of the pumpkings for it\,
independent of the sv\_2iv issue\.

Fix for sv\_2iv&#8203;:

Unfortunately\, looks\_like\_number has two functions&#8203;:
&nbsp;    \(a\) is it a number? and
&nbsp;    \(b\) what sort of number is it \(integer/rational\)?
looks\_like\_number has no concept of 'it looks most like an integer'\,
and reports 0 whatever's wrong\. Hence this patch assumes the worst case\.
It includes my previous patch at the end\, which can be snipped out if
not to taste\. I assume the dTHR is needed for the warning\.

<details><summary>Inline Patch</summary>

```diff
--- sv.c.orig	Mon Aug  2 22:29:12 1999
+++ sv.c	Thu Aug 19 23:26:25 1999
@@ -1156,4 +1156,11 @@
 	I32 numtype = looks_like_number(sv);
 
+	if (!numtype) {
+	    dTHR;
+	    if (ckWARN(WARN_NUMERIC))
+		not_a_number(sv);
+	    numtype = IS_NUMBER_NOT_IV;   /* Assume the worst case */
+	}
+
 	/* We want to avoid a possible problem when we cache an IV which
 	   may be later translated to an NV, and the resulting NV is not
@@ -1191,5 +1198,5 @@
 	    }
 	}
-	else if (numtype) {
+	else {
 	    /* The NV may be reconstructed from IV - safe to cache IV,
 	       which may be calculated by atol(). */
@@ -1199,14 +1206,4 @@
 	    SvIVX(sv) = Atol(SvPVX(sv));
 	}
-	else {				/* Not a number.  Cache 0. */
-	    dTHR;
-
-	    if (SvTYPE(sv) < SVt_PVIV)
-		sv_upgrade(sv, SVt_PVIV);
-	    SvIVX(sv) = 0;
-	    (void)SvIOK_on(sv);
-	    if (ckWARN(WARN_NUMERIC))
-		not_a_number(sv);
-	}
     }
     else  {
@@ -1635,5 +1632,5 @@
     while (isSPACE(*s))
 	s++;
-    if (s >= send)
+    if (s >= send || *s == 0)
 	return numtype;
     if (len == 10 && memEQ(sbegin, "0 but true", 10))
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu\, Aug 19\, 1999 at 11&#8203;:34&#8203;:32PM \+0100\, Ian Phillipps wrote&#8203;:
> Ilya > I would prefer sv\_2iv\(\) call conversion routine after emitting a
> Ilya > warning\.  \(I see no value in the testcase I sent \- one can always use
> Ilya > 'A' instead of 'a' in the unpack specifier\.\)
> 
> This patchlet was in support of Nick's agreement with my suggestion that
> "123\\0" should be treated as a fully legitimate number\. I still think
> it's Good Thing and crave the attention of the pumpkings for it\,
> independent of the sv\_2iv issue\.

I would prefer things to remain as they are\.  I saw no real\-line
complaints that looks\_like\_number\(\) is too picky\, and I can imagine a
lot of situations when "123\\0" is a result of a bug in a user code\.

Thus IMO there is no value in removing this strictness\, while there is
a value in preserving it\.

> Unfortunately\, looks\_like\_number has two functions&#8203;:
>     \(a\) is it a number? and
>     \(b\) what sort of number is it \(integer/rational\)?
> looks\_like\_number has no concept of 'it looks most like an integer'\,
> and reports 0 whatever's wrong\. 

But why this is an issue for sv\_2iv\(\) but not for sv\_2nv\(\)?

I find your argument not very convincing\.

Ilya
